atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/CompactPriorityCalculator.java

```diff
@@ -0,0 +1,110 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.atlasdb.transaction.api.Transaction;
+import com.palantir.atlasdb.transaction.api.TransactionManager;
+import com.palantir.logsafe.Arg;
+
+class CompactPriorityCalculator {
+    private static final Logger log = LoggerFactory.getLogger(CompactPriorityCalculator.class);
+
+    private final TransactionManager transactionManager;
+    private final CompactionHistoryProvider compactionHistoryProvider;
+    private final SweepHistoryProvider sweepHistoryProvider;
+
+    static CompactPriorityCalculator create(TransactionManager transactionManager) {
+        return new CompactPriorityCalculator(transactionManager,
+                new SweepHistoryProvider(),
+                new CompactionHistoryProvider());
+    }
+
+    @VisibleForTesting
+    CompactPriorityCalculator(TransactionManager transactionManager,
+            SweepHistoryProvider sweepHistoryProvider,
+            CompactionHistoryProvider compactionHistoryProvider) {
+        this.transactionManager = transactionManager;
+        this.sweepHistoryProvider = sweepHistoryProvider;
+        this.compactionHistoryProvider = compactionHistoryProvider;
+    }
+
+    Optional<String> selectTableToCompact() {
+        return transactionManager.runTaskReadOnly(this::selectTableToCompactInternal);
+    }
+
+    @VisibleForTesting
+    Optional<String> selectTableToCompactInternal(Transaction tx) {
+        Map<String, Long> tableToLastTimeSwept = sweepHistoryProvider.getHistory(tx);
+        Map<String, Long> tableToLastTimeCompacted = compactionHistoryProvider.getHistory(tx);
+
+        List<String> uncompactedTables = tableToLastTimeSwept.keySet().stream()
+                .filter(table -> !tableToLastTimeCompacted.keySet().contains(table))
+                .collect(Collectors.toList());
+
+        if (uncompactedTables.size() > 0) {
+            int randomTableIndex = ThreadLocalRandom.current().nextInt(uncompactedTables.size());
+            String randomlyChosenTable = uncompactedTables.get(randomTableIndex);
+            log.info("There are some tables which have been swept, but not compacted. Choosing {} at random.",
+                    safeTableRef(randomlyChosenTable));
+            return Optional.of(randomlyChosenTable);
+        }
+
+        String tableToCompact = null;
+        long maxSweptAfterCompact = Long.MIN_VALUE;
+        for (Map.Entry<String, Long> entry : tableToLastTimeSwept.entrySet()) {
+            String table = entry.getKey();
+            long lastSweptTime = entry.getValue();
+            long lastCompactTime = tableToLastTimeCompacted.get(table);
+            long sweptAfterCompact = lastSweptTime - lastCompactTime;
+
+            if (sweptAfterCompact > maxSweptAfterCompact) {
+                tableToCompact = table;
+                maxSweptAfterCompact = sweptAfterCompact;
+            }
+        }
```

Comment:
It feels like we can do something clever...

```
        tableToLastTimeSwept.entrySet().stream()
                .max(Comparator.comparingLong(entry -> entry.getValue() - tableToLastTimeCompacted.get(entry.getKey())))
                .map(Map.Entry::getKey);
```

and that returns an `Optional<String>`, but I'm probably getting ahead of myself here.

Reply:
Yep, saw that possibility but I kept the code I inherited unchanged for the sake of velocity.